### DISCUSSION TITLE
Is there a way to provide custom online help resulution for members of unsupported libraries?

### DISCUSSION THREAD

**Discussion by koszeggy** [EWSoftware/SHFB#991] [Q&A]
**Description:**

In some of my libraries I reference types in help pages that are not resolved automatically by SHFB.

Referencing such types (eg. [`SKBitmap`](https://learn.microsoft.com/en-us/dotnet/api/skiasharp.skbitmap) or WinUI's [`Color`](https://learn.microsoft.com/en-us/dotnet/api/windows.ui.color)) cause BE0066 warnings during the compilation.

For example:
```
SHFB(0,0): warning BE0066: ResolveReferenceLinksComponent: [M:KGySoft.Drawing.SkiaSharp.SKBitmapExtensions.GetReadWriteBitmapData(SkiaSharp.SKBitmap,SkiaSharp.SKColor,System.Byte)] Unknown reference link target 'T:SkiaSharp.SKBitmap'.
```

As a partial workaround now I replaced all `<see cref="..."/>` tags to `<a href="...">...<a>` in the documentation comments but it resolves only the links I explicitly put into the description, not the generated ones. So the generated help page still contains unresolved links and the compilation is full of BE0066 warnings I cannot avoid so it's really hard to spot the real issues that I should fix.

As an example, see [this help page](https://docs.kgysoft.net/drawing/html/M_KGySoft_Drawing_SkiaSharp_SKBitmapExtensions_ConvertPixelFormat.htm). I highlighted the result of the different references for the parameters:

![Help page issues](https://github.com/EWSoftware/SHFB/assets/27336165/dac6e4a8-ccdb-4d3d-8578-479f149080b6)
    <p align="center"><em>Annotated links of [this](https://docs.kgysoft.net/drawing/html/M_KGySoft_Drawing_SkiaSharp_SKBitmapExtensions_ConvertPixelFormat.htm) help page.</em></p>

* Green ellipses: Auto resolved links or explicit `<see cref="..."/>` references
* Red rectangles: Failed resolutions (`BE0066` warnings)
* Blue highlights: Explicit `<a href="...">...<a>` references as workaround where `<see cref="..."/>` references would not work (`BE0066`)

**The question:**

Is there a way to tell SHFB how to resolve these unknown targets? I would like to specify a custom mapping, something like this:
```
T:SkiaSharp.SKBitmap --> https://learn.microsoft.com/en-us/dotnet/api/skiasharp.skbitmap
T:Windows.UI.Color --> https://learn.microsoft.com/en-us/dotnet/api/windows.ui.color
etc.
```

Apparently neither the Additional Reference Links, nor the Member ID Fix-Ups plugins can be used for this (or just I couldn't figure it out how). It would be nice to allow any URL not just learn.microsoft.com if the documentation for a member is somewhere else.

**Comment by EWSoftware** [comment]
The Member ID Fix-Ups plug-in is for fixing issues with member IDs generated by the C++ and VB compilers.  The Additional Reference Links plug-in was mainly for the offline MSDN content.   For online content, it is generally used to suppress the warnings and generate non-clickable labels rather than links.

Out of the box there isn't a way to generate links to third party content but I plan on adding support at some point.  I just haven't had the time to look into it.  It has come up before and others have created plug-ins and components to handle it.  This is one I found from [issue 882](https://github.com/EWSoftware/SHFB/issues/882).  It may not be current and may require some updates to work with the latest release but appears to do what you want so it may be a good starting point.

**Reply by koszeggy** [reply]
Wow, thank you, I'm going to check it out.

**Comment by EWSoftware** [comment]
See #998 for an improved version of the noted plug-in.
